cpp/ql/lib/semmle/code/cpp/Variable.qll

```diff
@@ -169,6 +169,12 @@ class Variable extends Declaration, @variable {
     variable_instantiation(underlyingElement(this), unresolveElement(v))
   }
 
+  /**
+   * Holds if this variable is declated as part of a structured binding
+   * declaration. For example, `x` in `auto [x, y] = ...`.
+   */
+  predicate isStructuredBinding() { is_structured_binding(unresolveElement(this)) }
```

Comment:
I don't know much about the `unresolvelement` and `underlyingElement` (except for the fact that I know it's related to linker awareness). But all other predicates seem to use `underlyingElement(this)` instead of `unresolveElement(this)`. Is there a reason `unresolveElement(this)` is the correct thing to do here?

Reply:
I honestly just guessed based on what I saw happening in the surrounding code. Do we have any documentation on `unresolvElement` and `underlyingElement` explaining their meaning?

Reply:
There's _some_ documentation in the predicates defined near the top of `Element.qll`, but its far from great.

Reply:
Unfortunately, no. And I'm not sure that we'll be able to get proper documentation on these predicates since they precede most people on the C team (as can be witnessed by this very low PR number: https://github.com/github/codeql/pull/7). The best possible source of documentation is probably the QLDoc around these predicates: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/Element.qll#L44.

Reply:
Turns out `Element.qll` is very explicit on when to use `underlyingElement` and `unresolveElement`: use the former with `this` and the latter with anything that's not `this`. So this should be `underlyingElement`.

Reply:
Looking at the implementations that also makes sense to me, as `unresolveElement` does some special handling for incomplete classes, which is not something that is relevant here.

Reply:
Changed to `underlyingElement`. Thanks a lot for pointing this out!